Services/Storage/StorageRecords.cs

```diff
@@ -0,0 +1,371 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Concurrency;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.DataStructures;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Exceptions;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Storage.DocumentDb;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Storage
+{
+    public interface IStorageRecords
+    {
+        IStorageRecords Init(StorageConfig config);
+        Task<StorageRecord> GetAsync(string id);
+        Task<bool> ExistsAsync(string id);
+        Task<IEnumerable<StorageRecord>> GetAllAsync();
+        Task<StorageRecord> CreateAsync(StorageRecord input);
+        Task<StorageRecord> UpsertAsync(StorageRecord input);
+        Task<StorageRecord> UpsertAsync(StorageRecord input, string eTag);
+        Task DeleteAsync(string id);
+        Task DeleteMultiAsync(List<string> ids);
+        Task<bool> TryToLockAsync(string id, string ownerId, string ownerType, int durationSeconds);
+        Task<bool> TryToUnlockAsync(string id, string ownerId, string ownerType);
+    }
+
+    public class StorageRecords : IStorageRecords, IDisposable
+    {
+        private readonly ILogger log;
+        private readonly IInstance instance;
+        private readonly IDocumentDbWrapper docDb;
+        private readonly IConcurrencyConfig concurrencyConfig;
+        private StorageConfig storageConfig;
+
+        private IDocumentClient client;
+        private bool disposedValue;
+        private string storageName;
+
+        public StorageRecords(
+            IDocumentDbWrapper docDb,
+            ILogger logger,
+            IInstance instance,
+            IConcurrencyConfig concurrencyConfig)
+        {
+            this.log = logger;
+            this.instance = instance;
+            this.docDb = docDb;
+            this.disposedValue = false;
+            this.storageConfig = null;
+            this.client = null;
+            this.concurrencyConfig = concurrencyConfig;
```

Comment:
nit: move this to line 53, i.e. store the injected dependencies before initializing local fields

Reply:
Fixed, but I'd like to understand why this is important.

Reply:
it's not important, just a common aspect of our code base. All the ctors (should) store the dependencies first, and then do some work like initializing some local state. Note the "nit" in my comment, to convey that this is mostly about code style and readability, and optional, i.e. the code would work anyway.